docs/proposals/api-group.md

```diff
@@ -0,0 +1,104 @@
+
+1. No changes to API spec:
+
+  We will not add new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
+
+  For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid.
+
+2. conversion.Scheme:
+
+  The key of conversion.Scheme.versionMap will be "group/version". The internal version of objects of all groups will be registered to versionMap[""].
+
+3. meta.RESTMapper:
+
+  Each group will have its own RESTMapper, and these mappers will be registered to api.RESTMapper.
+
+  To support identically named kinds in different groups, We need to expand the input of RESTMapper.VersionAndKindForResource from (resource string) to (group, resource string). If group is not specified, the function should search groups in a predetermined order.
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+    * for api/v1, we set the prefix="api" (which is populated from app.APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1
+
+    * for new kube API groups, we will set the prefix="kube-api" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /kube-api/experimental/v1alpha1
+
+    * for OpenShift and other third party API, they should be able to name their own prefix and group. For example, OpenShift can either choose prefix="osapi", group="", or prefix="", group="osapi", so the URL remains the same as it is now.
+
+2. handling common API objects like api.Status:
+
+  To handle the API objects that are used by all groups, we either have to register them to all schemes, or we can choose not to encode them to a version.
+
+  Currently, for api.Status, we choose the latter approach: the server will encode it if the codec is v1, otherwise the server will send it unversioned.
+
+## Client-side implementation
+
+1. Client struct:
+
+  Currently each group has a client struct wrapping over \*RESTClient, e.g., client/unversioned.ExperimentalClient, client/unversioned.Client. The reason we have separate client struct is because each of them also implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). We don't know yet how to make their implementation scalable.
+
+2. Spelling the URL:
+
+  The URL is in the form of prefix/group/version/. The prefix is hard-coded in the client/unversioned.Config (see [here](../../pkg/client/unversioned/experimental.go#L101)). The client should be able to figure out `group` and `version` using the RESTMapper.
+
+3. kubectl:
+
+  kubectl shall accept arguments like `group/kind`, `group/kind/name`. Nevertheless, the user can omit the `group`, then kubectl shall rely on RESTMapper.VersionAndKindForResource() to find the most suitable group.
```

Comment:
Agreed, two restful representations of the same kind are both relevant and
valid targets to create.  We will always have cases where we need to POST
to a resource that creates another kind.

On Thu, Aug 20, 2015 at 3:16 PM, David Eads <notifications@github.com>
wrote:

> In docs/proposals/api-group.md
> <https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37569692>
> :
>
> > +
> > +  Currently, for api.Status, we choose the latter approach: the server will encode it if the codec is v1, otherwise the server will send it unversioned.
> > +
> > +## Client-side implementation
> > +
> > +1. Client struct:
> > +
> > +  Currently each group has a client struct wrapping over \*RESTClient, e.g., client/unversioned.ExperimentalClient, client/unversioned.Client. The reason we have separate client struct is because each of them also implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). We don't know yet how to make their implementation scalable.
> > +
> > +2. Spelling the URL:
> > +
> > +  The URL is in the form of prefix/group/version/. The prefix is hard-coded in the client/unversioned.Config (see [here](../../pkg/client/unversioned/experimental.go#L101)). The client should be able to figure out `group` and `version` using the RESTMapper.
> > +
> > +3. kubectl:
> > +
> > +  kubectl shall accept arguments like `group/kind`, `group/kind/name`. Nevertheless, the user can omit the `group`, then kubectl shall rely on RESTMapper.VersionAndKindForResource() to find the most suitable group.
>
> I want to be sure we're talking about the same thing. Today, kubectl
> accepts "resources", not "kinds". The distinction isn't important to users,
> but internally kinds and resources are not the same thing. I like the idea
> of having users specify resources. Today, there is a one-to-one mapping
> between kind and resource, but I'm not convinced that will always be the
> case. I'd like to make certain that this group/resource, not group/kind.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12951/files#r37569692>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift